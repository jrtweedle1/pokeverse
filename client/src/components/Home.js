import { useState, useEffect, createContext } from "react";
import Card from "./Card.js";

function Home() {
	const [allPokemon, setAllPokemon] = useState([]);
	const [squad, setSquad] = useState([]);

    // On page load, no more and no less, fetch all 151 original Pokemon
	useEffect(() => {
		fetchAllPokemon();
	}, []);

    // Initial function to populate all 151 original Pokemon
	async function fetchAllPokemon() {
		const initialPokemonJson = await fetch(
			"https://pokeapi.co/api/v2/pokemon?limit=151"
		);
		const initialPokemonNoResults = await initialPokemonJson.json();
		const initialPokemonResults = initialPokemonNoResults.results;
		const pokemonUrls = initialPokemonResults.map((pkmn) => pkmn.url);
		const pokemonDetails = await Promise.all(
			pokemonUrls.map(async (url) => {
				const response = await fetch(url);
				return await response.json();
			})
		);
		setAllPokemon(pokemonDetails);
	}

    // Adds Pokemon to squad, removes them from list of all Pokemon
	function addPokemonToSquad(singlePokemon) {
		const newAllPokemon = allPokemon.filter(function (pkmn) {
			return pkmn.name !== singlePokemon.name;
		});
		setAllPokemon(newAllPokemon);
		setSquad((prevSquad) => {
			return [...prevSquad, singlePokemon];
		});
	}

    // Removes Pokemon from squad, adds them to list of all Pokemon
	function removePokemonFromSquad(singlePokemon) {
		setSquad((squad) =>
			squad.filter(
				(squadPokemon) => squadPokemon.name !== singlePokemon.name
			)
		);
		setAllPokemon((currentAllPokemon) => [...allPokemon, singlePokemon]);
	}

    // Renders "homepage" in the DOM
	return (
        // Enables us to pass down squad info/functions to lower-level components
		<PokemonContext.Provider
			value={{ squad, addPokemonToSquad, removePokemonFromSquad }}
		>
			<div>
				<h1 className="text-center">Home</h1>
				<h2 className="text-center">My Squad</h2>
				<div className="container text-center">
					<div className="row">
						{squad.map((singlePokemon) => {
							return (
								<Card
									key={singlePokemon.name}
									pokemon={singlePokemon}
								></Card>
							);
						})}
					</div>
				</div>
				<h2 className="text-center">Available Pokemon</h2>
				<div className="container text-center">
					<div className="row">
						{allPokemon
							? allPokemon.map((singlePokemon) => {
									return (
										<Card
											key={singlePokemon.name}
											pokemon={singlePokemon}
										></Card>
									);
							  })
							: "Loading..."}
					</div>
				</div>
			</div>
		</PokemonContext.Provider>
	);
}

export default Home;
export const PokemonContext = createContext([]);
