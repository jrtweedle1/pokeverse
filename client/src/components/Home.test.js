import React from "react";
import { unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import renderer from "react-test-renderer";
import { TextEncoder, TextDecoder } from "util";
import { setupServer } from "msw/node";
import { rest } from "msw";
import { render, screen, waitFor } from "@testing-library/react";
import Home from "./Home.js";

// Declare container to which things will be rendered/mounted
let container = null;

// Declare encoders due to node version; necessary for tests to pass
global.TextEncoder = TextEncoder;
global.TextDecoder = TextDecoder;

// Setup server with MSW to mock API calls; these are called "handlers"
const server = setupServer(
    rest.get("https://pokeapi.co/api/v2/pokemon", (req, res, ctx) => {
        const limit = req.url.searchParams.get("limit");
        if (limit == "151") {
            return res(
                ctx.status(200),
                ctx.json({
                    results: [
                        {
                            name: "bulbasaur",
                            url: "https://pokeapi.co/api/v2/pokemon/1/",
                        },
                        {
                            name: "pikachu",
                            url: "https://pokeapi.co/api/v2/pokemon/25",
                        },
                    ],
                })
            );
        }
    }),
    rest.get("https://pokeapi.co/api/v2/pokemon/:pokemonId", (req, res, ctx) => {
        const { pokemonId } = req.params;
        if (pokemonId === "1") {
            return res(
                ctx.json({
                    id: 1,
                    name: "bulbasaur",
                    // Include more details here if I need for testing
                })
            );
        }
        if (pokemonId === "25") {
            return res(
                ctx.json({
                    id: 25,
                    name: "pikachu",
                    // Include more details here if I need for testing
                })
            );
        }
    })
);

// Establish API mocking before every test
beforeAll(() => server.listen());

// Sets up a clean environment before testing
beforeEach(() => {
	// Container is a DOM element that is a render target
	container = document.createElement("div");
	document.body.appendChild(container);
});

// Close the server and clean up after all tests
afterAll(() => server.close());

// Tears down environment after testing
afterEach(() => {
	// Unmount and "clean up" once test completes, in order to isolate the effects of a test to itself
	unmountComponentAtNode(container);
	container.remove();
	container = null;
	server.resetHandlers(); // Reset request handlers to prevent affecting other tests
});

// Snapshot - compares current version to previous snapshot
test("the component renders correctly", () => {
	const component = renderer.create(<Home />); // Creating an instance of the component
	const tree = component.toJSON(); // Convert to JSON
	expect(tree).toMatchSnapshot(); // Jest will create or update snapshot file
});

// Test API Call
test("displays fetched Pokemon on the homepage", async () => {
	render(<Home />);
	await waitFor(() => {
		expect(screen.getByText("bulbasaur")).toBeInTheDocument();
		expect(screen.getByText("pikachu")).toBeInTheDocument();
	});
});

// TODO: Test Component Rendering

// TODO: Test State Management
