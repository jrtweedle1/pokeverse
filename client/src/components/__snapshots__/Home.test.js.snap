// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the component renders correctly 1`] = `
<div>
  <h1
    className="text-center"
  >
    Home
  </h1>
  <h2
    className="text-center"
  >
    My Squad
  </h2>
  <div
    className="container text-center"
  >
    <div
      className="row"
    />
  </div>
  <h2
    className="text-center"
  >
    Available Pokemon
  </h2>
  <div
    className="container text-center"
  >
    <div
      className="row"
    />
  </div>
</div>
`;
